liota/core/metric_handler.py

```diff
@@ -30,4 +30,177 @@
+                    raise ValueError("'timeout' must be a non-negative number")
+                else:
+                    endtime = _time() + timeout
+                    while self._qsize() == self.maxsize:
+                        remaining = endtime - _time()
+                        if remaining <= 0.0:
+                            raise Full
+                        self.not_full.wait(remaining)
+            self._put(item)
+            self.unfinished_tasks += 1
+            self.not_empty.notify()
+
+            first_element_after_insertion = heapq.nsmallest(1, self.queue)[0]
+            if first_element_before_insertion != first_element_after_insertion:
+                self.first_element_changed.notify()
+        finally:
+            self.not_full.release()
+
+    def get_next_element_when_ready(self):
+        self.first_element_changed.acquire()
+        try:
+            isNotReady = True
+            while isNotReady:
+                if self._qsize() > 0:
+                    first_element = heapq.nsmallest(1, self.queue)[0]
+                    timeout = (first_element.get_next_run_time() -
+                               getUTCmillis()) / 1000.0
+                    log.info(
+                        "Waiting on acquired first_element_changed LOCK for: " + str(timeout))
+                    self.first_element_changed.wait(timeout)
+                else:
+                    self.first_element_changed.wait()
+                    first_element = heapq.nsmallest(1, self.queue)[0]
+                if (first_element.get_next_run_time() - getUTCmillis()) <= 0:
+                    isNotReady = False
+                    first_element = self._get()
+            return first_element
+        finally:
+            self.first_element_changed.release()
+
+
+class EventCheckerThread(Thread):
+
+    def __init__(self):
+        Thread.__init__(self)
+        self.start()
+
+    def run(self):
+        log.info("Started EventCheckerThread")
+        global event_ds
+        global collect_queue
+        while True:
+            log.debug("Waiting for event...")
+            metric = event_ds.get_next_element_when_ready()
+            log.debug("Got event:" + str(metric))
+            collect_queue.put(metric)
+
+
+class SendThread(Thread):
+
+    def __init__(self):
+        Thread.__init__(self)
+        self.start()
+
+    def run(self):
+        log.info("Started SendThread")
+        global send_queue
+        while True:
+            log.info("Waiting to send...")
+            metric = send_queue.get()
+            log.info("Got item in send_queue:" + str(metric))
+            metric.send_data()
+
+
+class CollectionThread(Thread):
+
+    def __init__(self):
+        Thread.__init__(self)
+        self.daemon = True
+        self.start()
+
+    def run(self):
+        global event_ds
+        global collect_queue
+        global send_queue
+        while True:
+            metric = collect_queue.get()
```

Comment:
Must check if the metric is marked as "deleted" (there should be a `flag_alive`).